oauthproxy.go

```diff
@@ -177,13 +179,14 @@ func NewOAuthProxy(opts *options.Options, validator func(string) bool) (*OAuthPr
 		CookieSameSite: opts.Cookie.SameSite,
 		Validator:      validator,
 
-		RobotsPath:        "/robots.txt",
-		SignInPath:        fmt.Sprintf("%s/sign_in", opts.ProxyPrefix),
-		SignOutPath:       fmt.Sprintf("%s/sign_out", opts.ProxyPrefix),
-		OAuthStartPath:    fmt.Sprintf("%s/start", opts.ProxyPrefix),
-		OAuthCallbackPath: fmt.Sprintf("%s/callback", opts.ProxyPrefix),
-		AuthOnlyPath:      fmt.Sprintf("%s/auth", opts.ProxyPrefix),
-		UserInfoPath:      fmt.Sprintf("%s/userinfo", opts.ProxyPrefix),
+		RobotsPath:              "/robots.txt",
+		SignInPath:              fmt.Sprintf("%s/sign_in", opts.ProxyPrefix),
+		SignOutPath:             fmt.Sprintf("%s/sign_out", opts.ProxyPrefix),
+		OAuthStartPath:          fmt.Sprintf("%s/start", opts.ProxyPrefix),
+		OAuthCallbackPath:       fmt.Sprintf("%s/callback", opts.ProxyPrefix),
+		AuthOnlyPath:            fmt.Sprintf("%s/auth", opts.ProxyPrefix),
+		UserInfoPath:            fmt.Sprintf("%s/userinfo", opts.ProxyPrefix),
+		AuthenticateOrStartPath: fmt.Sprintf("%s/auth_or_start", opts.ProxyPrefix), // originally contributed but not merged by: https://github.com/oauth2-proxy/oauth2-proxy/compare/master...maxlaverse:support_traefik?expand=1
```

Comment:
This path structure doesn't feel elegant. I feel like a querystring parameter that triggers a fallback to the `OAuthStart` handler instead of a `401 Unauthorized` in the failure case on `/auth` is more elegant.

Reply:
Seems pretty clean as well. Though again technically it isn't like `start` which immediately redirects, it's more like `sign_in` which either shows the various providers available or redirects when `--skip-provider-button=true`.

One downside to this approach vs having a new endpoint or new config option I can imagine, is that if someone assumes this functionality is available but is running an older version where it isn't *yet*, you wouldn't get any warnings/errors on startup. It would silently ignore the parameter and have unexpected behavior during actual requests that need authenticating.

In that regard I think either a config option or new endpoint is more predictable.

Reply:
Config endpoint would be "the best of both worlds" in this case, imho.
It would keep the endpoints elegant and keeps everything nice and tidy...

Reply:
Is the intent to fallback to sign-in or start page? I'm tempted to lean toward `start` to align with recommendations for other `subrequest` recommended architectures - `SignInPage` kills existing OAuth2-Proxy session cookies.

But I could be swayed, I see the utility if people like the sign-in page (I hate it, get me to my IdP 😄 )

Could be chosen based on the option querystring param proposal? `?unauthorized=(?:start|sign_in)`

Reply:
> One downside to this approach vs having a new endpoint or new config option I can imagine, is that if someone assumes this functionality is available but is running an older version where it isn't _yet_, you wouldn't get any warnings/errors on startup. It would silently ignore the parameter and have unexpected behavior during actual requests that need authenticating.

This is a known issue in many areas we hope to address with restructuring our docs.  With the current docs representing master, we frequently get users confused by following the for configuration options that have merges but aren't in a release yet.

So I wouldn't let this be concern for our design decision.

Reply:
> Config option for the auth endpoint would be "the best of both worlds" in this case, imho.
> It would keep the endpoints elegant and keeps everything nice and tidy...

Didn't read this closely enough -- Agreed! Config option is best to give an alternative to `/oauth2/auth` failure behavior